// @midwayjs/runtime-engine/dist/lib/loggerFactory.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.BaseLoggerFactory = void 0;
const path_1 = require("path");
const logger_1 = require("./logger");
class BaseLoggerFactory {
    constructor(homeDir, envParser, Logger) {
        this.homeDir = homeDir;
        this.baseDir = (0, path_1.join)(this.homeDir, 'logs');
        this.envParser = envParser;
        this.Logger = Logger || logger_1.ServerlessLogger;
    }
    createLogger(filename, options) {
        if (typeof filename === 'string') {
            options = options || {};
            options.file = filename;
        }
        else {
            options = filename || {};
        }
        const lv = this.envParser ? this.envParser.getLoggerLevel() : 'ERROR';
        if (lv) {
            options.level = lv;
            options.consoleLevel = lv;
        }
        const Logger = this.Logger;
        const logger = new Logger(Object.assign({}, options, {
            file: options.file ? (0, path_1.join)(this.baseDir, options.file) : undefined,
            baseDir: this.baseDir,
        }));
        return logger;
    }
}
exports.BaseLoggerFactory = BaseLoggerFactory;
