// @midwayjs/runtime-engine/dist/lib/parser.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.EnvPropertyParser = exports.EnvPropertyKey = void 0;
var EnvPropertyKey;
(function (EnvPropertyKey) {
    EnvPropertyKey["INIT_TIMEOUT"] = "INIT_TIMEOUT";
    EnvPropertyKey["INIT_HANDLER"] = "INIT_HANDLER";
    EnvPropertyKey["FUNC_TIMEOUT"] = "FUNC_TIMEOUT";
    EnvPropertyKey["FUNC_HANDLER"] = "FUNC_HANDLER";
    EnvPropertyKey["FUNC_RUNTIME"] = "FUNC_RUNTIME";
    EnvPropertyKey["ENTRY_DIR"] = "ENTRY_DIR";
    EnvPropertyKey["TRIGGER_TYPE"] = "TRIGGER_TYPE";
    EnvPropertyKey["FUNC_LAYER"] = "FUNC_LAYER";
    EnvPropertyKey["LOGER_LEVEL"] = "LOGGER_LEVEL";
    EnvPropertyKey["EAGLEEYE_FLAG"] = "EAGLE_FLAG";
})(EnvPropertyKey = exports.EnvPropertyKey || (exports.EnvPropertyKey = {}));
class EnvPropertyParser extends Map {
    setProperty(propertyKey, value) {
        process.env[propertyKey] = value;
    }
    getProperty(propertyKey, defaultValue) {
        if (!this.has(propertyKey)) {
            if (process.env[propertyKey]) {
                this.set(propertyKey, process.env[propertyKey]);
            }
            else if (process.env[propertyKey.toUpperCase()]) {
                this.set(propertyKey, process.env[propertyKey.toUpperCase()]);
            }
            else if (process.env[propertyKey.toLowerCase()]) {
                this.set(propertyKey, process.env[propertyKey.toLowerCase()]);
            }
            else {
                this.set(propertyKey, defaultValue || '');
            }
        }
        return this.get(propertyKey);
    }
    getInitTimeout() {
        return this.getProperty(EnvPropertyKey.INIT_TIMEOUT, 3 * 60 * 1000); // 3min
    }
    getFuncTimeout() {
        return this.getProperty(EnvPropertyKey.FUNC_TIMEOUT, 3 * 60 * 1000); // 3min
    }
    getInitHandler() {
        const funcHandler = this.getFunctionHandler();
        const [entryFile] = funcHandler.split('.');
        return this.getProperty(EnvPropertyKey.INIT_HANDLER, `${entryFile}.initialize`);
    }
    getFunctionHandler() {
        return this.getProperty(EnvPropertyKey.FUNC_HANDLER, 'index.handler');
    }
    getFunctionRuntime() {
        return this.getProperty(EnvPropertyKey.FUNC_RUNTIME, 'function_runtime');
    }
    getEntryDir() {
        return this.getProperty(EnvPropertyKey.ENTRY_DIR, process.cwd());
    }
    getTriggerType() {
        return this.getProperty(EnvPropertyKey.TRIGGER_TYPE, 'default_trigger');
    }
    getFunctionLayer() {
        return this.getProperty(EnvPropertyKey.FUNC_LAYER, '');
    }
    getLoggerLevel() {
        return this.getProperty(EnvPropertyKey.LOGER_LEVEL, 'WARN');
    }
    getTraceFlag() {
        return this.getProperty(EnvPropertyKey.EAGLEEYE_FLAG, 'on');
    }
}
exports.EnvPropertyParser = EnvPropertyParser;
