// light-generator/dist/generator/CommonGenerator.js
"use strict";
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.CommonGenerator = void 0;
const interface_1 = require("../interface");
const path_1 = require("path");
const fs_1 = require("../util/fs");
const fs_extra_1 = require("fs-extra");
const untildify_1 = __importDefault(require("untildify"));
const empty_dir_1 = __importDefault(require("empty-dir"));
class CommonGenerator {
    constructor(options) {
        this.templateUri = options.templateUri;
        this.targetPath = options.targetPath;
        this.copyWalker = options.copyWalker;
        this.templateName = options.templateName;
        this.eventCenter = options.eventCenter;
    }
    /**
     * 获取模板配置
     */
    async getTemplateConfig() {
        if (!this.templateConfig) {
            const templateRoot = this.getTemplatePath();
            const templatePkg = path_1.join(templateRoot, 'package.json');
            if (fs_1.fileExistsSync(templatePkg)) {
                try {
                    const pkg = fs_1.readFileSync(templatePkg);
                    const config = pkg['boilerplateConfig'];
                    if (config) {
                        config.replaceFile = config.replaceFile || ['README.md'];
                        if (typeof config.replaceFile === 'string') {
                            config.replaceFile = [config.replaceFile];
                        }
                        // normalize path for windows
                        config.replaceFile = config.replaceFile.map(item => {
                            return path_1.normalize(item);
                        });
                        config.replaceParameter =
                            config.replaceParameter || 'index.js';
                        config.root = path_1.join(templateRoot, config.root || 'boilerplate');
                        config.rule = config.rule || [];
                        this.templateConfig = config;
                    }
                }
                catch (err) {
                    console.log('boilerplate package.json is not custom format and skip');
                }
            }
        }
        return this.templateConfig;
    }
    /**
     * 获取参数列表
     */
    async getParameterList() {
        const templateConfig = (await this.getTemplateConfig());
        const templateRoot = this.getTemplatePath();
        if (templateConfig) {
            const parameterFile = path_1.join(templateRoot, templateConfig.replaceParameter);
            if (fs_1.fileExistsSync(parameterFile)) {
                return require(parameterFile);
            }
        }
        return {};
    }
    /**
     * 获取参数默认值
     */
    async getDefaultParameterValue() {
        const parameterList = await this.getParameterList();
        const defaultValue = {};
        for (const key in parameterList) {
            if (parameterList[key]['default']) {
                defaultValue[key] = parameterList[key]['default'];
            }
        }
        return defaultValue;
    }
    async run(replaceParameter = {}) {
        // Copying template from a local directory
        const servicePath = untildify_1.default(this.targetPath);
        if (fs_1.dirExistsSync(servicePath)) {
            const checkResult = await empty_dir_1.default(servicePath);
            if (!checkResult) {
                const errorMessage = `A folder named "${servicePath}" already exists.`;
                throw new Error(errorMessage);
            }
        }
        else {
            // create
            await fs_extra_1.ensureDir(servicePath);
        }
        let templateConfig = (await this.getTemplateConfig());
        let templateRoot = this.getTemplatePath();
        const packageRoot = templateRoot;
        if (templateConfig) {
            templateRoot = templateConfig.root;
            if (!fs_1.dirExistsSync(templateRoot)) {
                throw new Error(`Directory ${templateRoot} not exist`);
            }
        }
        else {
            templateConfig = {
                replaceParameter: 'index.js',
                root: templateRoot,
                replaceFile: ['README.md'],
                rule: [],
            };
        }
        this.eventCenter.emit(interface_1.GeneratorEventEnum.onTemplateReady);
        const defaultArgsValue = await this.getDefaultParameterValue();
        replaceParameter = Object.assign(defaultArgsValue, replaceParameter);
        if (templateConfig.beforeAll) {
            await this.runScript(packageRoot, templateConfig.beforeAll, {
                sourceRoot: packageRoot,
                templateRoot,
                targetRoot: servicePath,
                replaceParameter,
                templateConfig,
            });
        }
        await this.copyWalker.copy(templateRoot, servicePath, {
            packageRoot,
            replaceParameter,
            templateConfig,
            noLinks: true,
        });
        if (templateConfig.afterAll) {
            await this.runScript(packageRoot, templateConfig.afterAll, {
                sourceRoot: packageRoot,
                templateRoot,
                targetRoot: servicePath,
                replaceParameter,
                templateConfig,
            });
        }
    }
    async runScript(packageRoot, runString, runArgs) {
        const fn = path_1.isAbsolute(runString)
            ? require(runString)
            : require(path_1.join(packageRoot, runString));
        if (fn && typeof fn === 'function') {
            await fn(runArgs);
        }
    }
    onTemplateReady(handler) {
        this.eventCenter.on(interface_1.GeneratorEventEnum.onTemplateReady, handler);
    }
    onFileCreated(handler) {
        this.eventCenter.on(interface_1.GeneratorEventEnum.onFileCreate, handler);
    }
}
exports.CommonGenerator = CommonGenerator;
