// light-generator/dist/generator/NpmPatternGenerator.js
"use strict";
var __createBinding = (this && this.__createBinding) || (Object.create ? (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    Object.defineProperty(o, k2, { enumerable: true, get: function() { return m[k]; } });
}) : (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    o[k2] = m[k];
}));
var __setModuleDefault = (this && this.__setModuleDefault) || (Object.create ? (function(o, v) {
    Object.defineProperty(o, "default", { enumerable: true, value: v });
}) : function(o, v) {
    o["default"] = v;
});
var __importStar = (this && this.__importStar) || function (mod) {
    if (mod && mod.__esModule) return mod;
    var result = {};
    if (mod != null) for (var k in mod) if (k !== "default" && Object.prototype.hasOwnProperty.call(mod, k)) __createBinding(result, mod, k);
    __setModuleDefault(result, mod);
    return result;
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.NpmPatternGenerator = void 0;
const CommonGenerator_1 = require("./CommonGenerator");
const child_process_1 = require("child_process");
const path_1 = require("path");
const fse = __importStar(require("fs-extra"));
const fs_1 = require("../util/fs");
const tar = __importStar(require("tar"));
const util_1 = require("../util/");
const util_2 = require("util");
const debugLogger = util_2.debuglog('generator:npm');
class NpmPatternGenerator extends CommonGenerator_1.CommonGenerator {
    constructor(options) {
        super(options);
        this.npmClient = options.npmClient;
        this.npmInstall = !!options.npmInstall;
        this.registryUrl = options.registryUrl
            ? '--registry=' + options.registryUrl
            : '';
        this.tmpPath = util_1.getTmpDir();
        this.targetVersion = options.targetVersion;
        fse.ensureDirSync(this.tmpPath);
        debugLogger('current npm module = [%s]', this.npmClient);
    }
    async getPackage() {
        const data = child_process_1.execSync(`${this.npmClient} view ${this.templateUri} dist-tags --json ${this.registryUrl}`, {
            cwd: process.env.HOME,
        }).toString();
        const remoteVersion = JSON.parse(data)[this.targetVersion || 'latest'];
        this.pkgRootName = `${util_1.renamePackageName(this.templateUri)}-${remoteVersion}`;
        const currentPkgRoot = this.getTemplatePath();
        debugLogger('currentPkgRoot = [%s], tmpPath = [%s]', currentPkgRoot, this.tmpPath);
        // 清理失败的模板
        if (fs_1.dirExistsSync(currentPkgRoot) &&
            !fs_1.fileExistsSync(path_1.join(currentPkgRoot, '.success'))) {
            await fse.remove(currentPkgRoot);
        }
        if (!fs_1.dirExistsSync(currentPkgRoot)) {
            // clean template directory first
            if (fs_1.dirExistsSync(path_1.join(this.tmpPath, this.pkgRootName))) {
                await fse.remove(path_1.join(this.tmpPath, this.pkgRootName));
            }
            const cmd = `${this.npmClient} pack ${this.templateUri}@${remoteVersion} ${this.registryUrl}&& mkdir ${this.pkgRootName}`;
            debugLogger('download cmd = [%s]', cmd);
            // run download
            child_process_1.execSync(cmd, {
                cwd: this.tmpPath,
                stdio: ['pipe', 'ignore', 'pipe'],
            });
            await tar.x({
                file: path_1.join(this.tmpPath, `${this.pkgRootName}.tgz`),
                C: path_1.join(this.tmpPath, this.pkgRootName),
            });
            if (!fs_1.dirExistsSync(currentPkgRoot)) {
                throw new Error(`${currentPkgRoot} package download error`);
            }
            if (this.npmInstall && fse.existsSync(path_1.join(currentPkgRoot, 'package.json'))) {
                const pkg = require(path_1.join(currentPkgRoot, 'package.json'));
                if (pkg['dependencies']) {
                    debugLogger('find package.json and dependencies');
                    const installCmd = `${this.npmClient} ${this.registryUrl} install --production`;
                    child_process_1.execSync(installCmd, {
                        cwd: currentPkgRoot,
                        stdio: ['pipe', 'ignore', 'pipe'],
                    });
                    debugLogger('install dependencies complete');
                }
            }
            await fse.writeFile(path_1.join(currentPkgRoot, '.success'), 'complete');
        }
    }
    async getTemplateConfig() {
        await this.getPackage();
        return super.getTemplateConfig();
    }
    getTemplatePath() {
        return path_1.join(this.tmpPath, this.pkgRootName, 'package');
    }
}
exports.NpmPatternGenerator = NpmPatternGenerator;
