// light-generator/dist/generator.js
"use strict";
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.LightGenerator = void 0;
const copyDirContents_1 = require("./util/copyDirContents");
const NpmPatternGenerator_1 = require("./generator/NpmPatternGenerator");
const LocalPatternGenerator_1 = require("./generator/LocalPatternGenerator");
const rule_1 = require("./rule");
const util_1 = require("./util/");
const fs_1 = require("./util/fs");
const fs_extra_1 = require("fs-extra");
const events_1 = __importDefault(require("events"));
class LightGenerator {
    constructor(options = { disableDefaultRule: false }) {
        this.options = options;
        this.eventCenter = new events_1.default();
        this.copyWalker = new copyDirContents_1.DirectoryCopyWalker(Object.assign(this.options, {
            eventCenter: this.eventCenter,
        }));
        if (!this.options.disableDefaultRule) {
            this.addDefaultCopyRule();
        }
    }
    addDefaultCopyRule() {
        this.copyWalker.addCopyRule(rule_1.replaceRule);
        this.copyWalker.addCopyRule(rule_1.ignoreRule);
    }
    defineLocalPath(options) {
        return new LocalPatternGenerator_1.LocalPatternGenerator({
            templateUri: options.templatePath,
            targetPath: options.targetPath,
            templateName: options.templateName,
            copyWalker: this.copyWalker,
            eventCenter: this.eventCenter,
        });
    }
    defineNpmPackage(options) {
        let npmClient = options.npmClient || 'npm';
        // 内部执行的命令，yarn 没有，所以固化为 npm
        if (npmClient === 'yarn') {
            npmClient = 'npm';
        }
        // 默认安装，可选择跳过
        const npmInstall = typeof options.npmInstall === 'boolean' ? options.npmInstall : true;
        return new NpmPatternGenerator_1.NpmPatternGenerator({
            templateUri: options.npmPackage,
            targetPath: options.targetPath,
            copyWalker: this.copyWalker,
            npmClient,
            registryUrl: options.registryUrl,
            eventCenter: this.eventCenter,
            npmInstall,
            targetVersion: options.targetVersion || 'latest',
        });
    }
    static async cleanCache() {
        const tmpDir = util_1.getTmpDir();
        if (fs_1.dirExistsSync(tmpDir)) {
            await fs_extra_1.remove(tmpDir);
        }
    }
}
exports.LightGenerator = LightGenerator;
