// @midwayjs/logger/dist/contextLogger.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.MidwayContextLogger = void 0;
class MidwayContextLogger {
    constructor(ctx, contextLogger) {
        this.ctx = ctx;
        this.contextLogger = contextLogger;
    }
    log(...args) {
        if (!['debug', 'info', 'warn', 'error'].includes(args[0])) {
            args.unshift('info');
        }
        this.transformLog('log', args);
    }
    debug(...args) {
        this.transformLog('debug', args);
    }
    info(...args) {
        this.transformLog('info', args);
    }
    warn(...args) {
        this.transformLog('warn', args);
    }
    error(...args) {
        this.transformLog('error', args);
    }
    transformLog(level, args) {
        return this.contextLogger[level].apply(this.contextLogger, [
            ...args,
            {
                label: this.formatContextLabel(),
                ctx: this.ctx,
            },
        ]);
    }
    formatContextLabel() {
        return '';
    }
}
exports.MidwayContextLogger = MidwayContextLogger;
