// @midwayjs/runtime-engine/dist/bootstrap.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.BaseBootstrap = void 0;
const engine_1 = require("./engine");
class BaseBootstrap {
    constructor(options = {}) {
        this.options = options;
        this.runtime = this.options.runtime;
        this.runtimeEngine = new engine_1.BaseRuntimeEngine();
        this.runtimeEngine.add(engine => {
            engine.addBaseRuntime(options.runtime);
        });
        // set options
        if (this.runtime) {
            this.runtime.setOptions(this.options);
        }
    }
    async start() {
        if (this.options.layers && this.options.layers.length) {
            this.options.layers.map(mod => this.runtimeEngine.add(mod));
        }
        await this.runtimeEngine.ready();
        return this.runtimeEngine.getCurrentRuntime();
    }
    async close() {
        return this.runtimeEngine.close();
    }
    getRuntime() {
        return this.runtimeEngine.getCurrentRuntime();
    }
    getRuntimeEngine() {
        return this.runtimeEngine;
    }
}
exports.BaseBootstrap = BaseBootstrap;
