// @midwayjs/serverless-fc-trigger/dist/cdn.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.createCDNEvent = exports.cdn = exports.CDNTrigger = void 0;
const base_1 = require("./base");
const extend = require("extend2");
/**
 * https://help.aliyun.com/document_detail/62922.html
 */
class CDNTrigger extends base_1.FCBaseTrigger {
    getEvent() {
        return {
            events: [
                {
                    eventName: 'LogFileCreated',
                    eventSource: 'cdn',
                    region: 'cn-hangzhou',
                    eventVersion: '1.0.0',
                    eventTime: 'xxxxx',
                    userIdentity: {
                        aliUid: '1xxxxxxxxxxxx',
                    },
                    resource: {
                        domain: 'example.com',
                    },
                    eventParameter: {
                        domain: 'example.com',
                        endTime: 1000000,
                        fileSize: 1788115,
                        filePath: 'http://cdnlog.cn-hangzhou.oss.aliyun-inc.com/xxx/xxx/xxx.gz?OSSAccessKeyId=xxxx&Expires=xxxx&Signature=xxxx',
                        startTime: 11000000,
                    },
                    traceId: 'xxxxxx',
                },
            ],
        };
    }
}
exports.CDNTrigger = CDNTrigger;
exports.cdn = CDNTrigger;
const createCDNEvent = (data = {}) => {
    return extend(true, new CDNTrigger().getEvent(), data);
};
exports.createCDNEvent = createCDNEvent;
