// @midwayjs/serverless-fc-trigger/dist/sls.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.createSLSEvent = exports.sls = exports.SLSTrigger = void 0;
const base_1 = require("./base");
const extend = require("extend2");
/**
 * https://help.aliyun.com/document_detail/84092.html
 */
class SLSTrigger extends base_1.FCBaseTrigger {
    getEvent() {
        return {
            parameter: {},
            source: {
                endpoint: 'http://cn-shanghai-intranet.log.aliyuncs.com',
                projectName: 'log-com',
                logstoreName: 'log-en',
                shardId: 0,
                beginCursor: 'MTUyOTQ4MDIwOTY1NTk3ODQ2Mw==',
                endCursor: 'MTUyOTQ4MDIwOTY1NTk3ODQ2NA==',
            },
            jobName: '1f7043ced683de1a4e3d8d70b5a412843d817a39',
            taskId: 'c2691505-38da-4d1b-998a-f1d4bb8c9994',
            cursorTime: 1529486425,
        };
    }
}
exports.SLSTrigger = SLSTrigger;
exports.sls = SLSTrigger;
const createSLSEvent = (data = {}) => {
    return extend(true, new SLSTrigger().getEvent(), data);
};
exports.createSLSEvent = createSLSEvent;
