// @midwayjs/serverless-fc-trigger/dist/tableStore.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.createTableStoreEvent = exports.tableStore = exports.TableStoreTrigger = void 0;
const base_1 = require("./base");
const extend = require("extend2");
/**
 * https://help.aliyun.com/document_detail/100092.html
 */
class TableStoreTrigger extends base_1.FCBaseTrigger {
    getEvent() {
        return {
            Version: 'string',
            Records: [
                {
                    Type: 'string',
                    Info: {
                        Timestamp: Date.now(),
                    },
                    PrimaryKey: [
                        {
                            ColumnName: 'string',
                            Value: 'new event',
                        },
                    ],
                    Columns: [
                        {
                            Type: 'string',
                            ColumnName: 'string',
                            Value: 'new event',
                            Timestamp: Date.now(),
                        },
                    ],
                },
            ],
        };
    }
}
exports.TableStoreTrigger = TableStoreTrigger;
exports.tableStore = TableStoreTrigger;
const createTableStoreEvent = (data = {}) => {
    return extend(true, new TableStoreTrigger().getEvent(), data);
};
exports.createTableStoreEvent = createTableStoreEvent;
