// light-generator/dist/util/copyDirContents.js
"use strict";
var __createBinding = (this && this.__createBinding) || (Object.create ? (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    Object.defineProperty(o, k2, { enumerable: true, get: function() { return m[k]; } });
}) : (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    o[k2] = m[k];
}));
var __setModuleDefault = (this && this.__setModuleDefault) || (Object.create ? (function(o, v) {
    Object.defineProperty(o, "default", { enumerable: true, value: v });
}) : function(o, v) {
    o["default"] = v;
});
var __importStar = (this && this.__importStar) || function (mod) {
    if (mod && mod.__esModule) return mod;
    var result = {};
    if (mod != null) for (var k in mod) if (k !== "default" && Object.prototype.hasOwnProperty.call(mod, k)) __createBinding(result, mod, k);
    __setModuleDefault(result, mod);
    return result;
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.DirectoryCopyWalker = void 0;
const path = __importStar(require("path"));
const fse = __importStar(require("fs-extra"));
const walkDirSync_1 = require("./walkDirSync");
const interface_1 = require("../interface");
const debug = require('util').debuglog('generator');
class DirectoryCopyWalker {
    constructor(options = {}) {
        this.rules = options.rules || [];
        this.eventCenter = options.eventCenter;
    }
    addCopyRule(rule) {
        this.rules.push(rule);
    }
    async copy(srcDir, destDir, options = {
        replaceParameter: {},
        templateConfig: {},
    }) {
        const fullFilesPaths = walkDirSync_1.walkDirSync(srcDir, options);
        const filenameMapping = new Map();
        // add custom rule
        if (options.templateConfig.rule && options.templateConfig.rule.length) {
            for (const rule of options.templateConfig.rule) {
                try {
                    const copyRule = path.isAbsolute(rule)
                        ? require(rule)
                        : require(path.join(options.packageRoot, rule));
                    this.addCopyRule(copyRule);
                }
                catch (err) {
                    throw new Error(`load custom rule error, path = ${rule}`);
                }
            }
        }
        for (const fullFilePath of fullFilesPaths) {
            const relativeFilePath = path.relative(srcDir, fullFilePath);
            const targetFilePath = path.join(destDir, relativeFilePath);
            await fse.copy(fullFilePath, path.join(destDir, relativeFilePath));
            this.eventCenter.emit(interface_1.GeneratorEventEnum.onFileCreate, {
                sourceFullFilePath: fullFilePath,
                targetFullFilePath: path.join(destDir, relativeFilePath),
                destDir,
                relativeFilePath,
            });
            debug(`-> ${relativeFilePath}`);
            for (const rule of this.rules) {
                await rule(targetFilePath, {
                    templateDir: srcDir,
                    targetDir: destDir,
                    replaceParameter: options.replaceParameter,
                    targetRelativeFile: relativeFilePath,
                    templateConfig: options.templateConfig || {
                        root: destDir,
                        replaceFile: [],
                        replaceParameter: {},
                        rule: [],
                    },
                    filenameMapping,
                });
            }
        }
    }
}
exports.DirectoryCopyWalker = DirectoryCopyWalker;
