// @midwayjs/serverless-app/dist/start.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.start2 = void 0;
const { BootstrapStarter } = require('@midwayjs/bootstrap');
const utils_1 = require("./utils");
const start2 = async (options) => {
    const { appDir, baseDir, framework, starter, layers = [], initializeContext, preloadModules, } = options;
    const { start } = starter;
    let starterInstance;
    let boot;
    layers.unshift(engine => {
        engine.addRuntimeExtension({
            async beforeFunctionStart(runtime) {
                starterInstance = new framework();
                starterInstance.configure({
                    initializeContext,
                    preloadModules,
                    applicationAdapter: runtime,
                });
                boot = new BootstrapStarter();
                boot
                    .configure({
                    baseDir,
                })
                    .load(starterInstance);
                await boot.init();
                await boot.run();
            },
        });
    });
    const runtime = await start({
        layers: layers,
        getApp: () => {
            return starterInstance && starterInstance.getApplication();
        },
        initContext: initializeContext,
    });
    return {
        runtime,
        framework: starterInstance,
        innerBootStarter: boot,
        // 分析装饰器上面的函数表
        getFunctionsFromDecorator: async () => {
            return (0, utils_1.analysisDecorator)(appDir);
        },
        invoke: async (handlerName, trigger) => {
            return runtime.asyncEvent(async (...args) => {
                return starterInstance.handleInvokeWrapper(handlerName)(...args);
            })(...trigger);
        },
    };
};
exports.start2 = start2;
