// @midwayjs/mock/dist/index.js
"use strict";
var __createBinding = (this && this.__createBinding) || (Object.create ? (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    Object.defineProperty(o, k2, { enumerable: true, get: function() { return m[k]; } });
}) : (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    o[k2] = m[k];
}));
var __exportStar = (this && this.__exportStar) || function(m, exports) {
    for (var p in m) if (p !== "default" && !Object.prototype.hasOwnProperty.call(exports, p)) __createBinding(exports, m, p);
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.createLightApp = exports.createBootstrap = exports.createFunctionApp = exports.createApp = exports.close = exports.create = void 0;
var utils_1 = require("./utils");
Object.defineProperty(exports, "create", { enumerable: true, get: function () { return utils_1.create; } });
Object.defineProperty(exports, "close", { enumerable: true, get: function () { return utils_1.close; } });
Object.defineProperty(exports, "createApp", { enumerable: true, get: function () { return utils_1.createApp; } });
Object.defineProperty(exports, "createFunctionApp", { enumerable: true, get: function () { return utils_1.createFunctionApp; } });
Object.defineProperty(exports, "createBootstrap", { enumerable: true, get: function () { return utils_1.createBootstrap; } });
Object.defineProperty(exports, "createLightApp", { enumerable: true, get: function () { return utils_1.createLightApp; } });
__exportStar(require("./client/http"), exports);
__exportStar(require("./client/rabbitMQ"), exports);
__exportStar(require("./client/socketio"), exports);
__exportStar(require("./client/ws"), exports);
/**
 * @deprecated below
 */
__exportStar(require("egg-mock"), exports);
__exportStar(require("./interface"), exports);
